atproto/repo/mst/operation.go

```diff
@@ -0,0 +1,150 @@
+package mst
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/ipfs/go-cid"
+)
+
+type Operation struct {
+	Path  string
+	Value *cid.Cid
+	Prev  *cid.Cid
+}
+
+func (op *Operation) IsCreate() bool {
+	if op.Value != nil && op.Prev == nil {
+		return true
+	}
+	return false
+}
+
+func (op *Operation) IsUpdate() bool {
+	if op.Value != nil && op.Prev != nil && *op.Value != *op.Prev {
+		return true
+	}
+	return false
+}
+
+func (op *Operation) IsDelete() bool {
+	if op.Value == nil && op.Prev != nil {
+		return true
+	}
+	return false
+}
+
+// Mutates the tree, returning a full `Operation`
+func ApplyOp(tree *Tree, path string, val *cid.Cid) (*Operation, error) {
```

Comment:
why not member of Tree? `someTree.ApplyOp(path, val)` ?

Reply:
Why isn't this two functions? `someTree.InsertCid(val)` and `someTree.RemovePath(path)`

Reply:
The Operation code is a bit funky because it is sort of a separate type/concept from the MST. maybe this code belongs better in the `atproto/repo` package? This is why I didn't want to put the methods on `Tree` directly; they don't feel like part of the `Tree` API.

I think it is helpful to keep the `Operation` abstraction around distinct from the lower-level insert/remove methods. This is all clearer at the `Repo` abstraction layer, so i'll move the code up to there.